libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -37,50 +42,50 @@ class ordered_diff {
       diff_result result;
       while (s < source.size() || t < target.size()) {
          if (s < source.size() && t < target.size()) {
+            FC_ASSERT(s < std::numeric_limits<SizeType>::max());
+            FC_ASSERT(t < std::numeric_limits<SizeType>::max());
             if (source[s] == target[t]) {
                // nothing to do, skip over
-               assert(s <= std::numeric_limits<SizeType>::max());
-               assert(t <= std::numeric_limits<SizeType>::max());
                ++s;
                ++t;
             } else { // not equal
                if (s == source.size() - 1 && t == target.size() - 1) {
                   // both at end, insert target and remove source
-                  assert(s <= std::numeric_limits<SizeType>::max());
-                  assert(t <= std::numeric_limits<SizeType>::max());
                   result.remove_indexes.push_back(s);
                   result.insert_indexes.emplace_back(t, target[t]);
+                  FC_ASSERT(result.remove_indexes.size() <= MAX_NUM_ARRAY_ELEMENTS);
+                  FC_ASSERT(result.insert_indexes.size() <= MAX_NUM_ARRAY_ELEMENTS);
                   ++s;
                   ++t;
                } else if (s + 1 < source.size() && t + 1 < target.size() && source[s + 1] == target[t + 1]) {
                   // misalignment, but next value equal, insert and remove
-                  assert(s <= std::numeric_limits<SizeType>::max());
-                  assert(t <= std::numeric_limits<SizeType>::max());
                   result.remove_indexes.push_back(s);
                   result.insert_indexes.emplace_back(t, target[t]);
+                  FC_ASSERT(result.remove_indexes.size() <= MAX_NUM_ARRAY_ELEMENTS);
+                  FC_ASSERT(result.insert_indexes.size() <= MAX_NUM_ARRAY_ELEMENTS);
                   ++s;
                   ++t;
                } else if (t + 1 < target.size() && source[s] == target[t + 1]) {
                   // source equals next target, insert current target
-                  assert(t <= std::numeric_limits<SizeType>::max());
                   result.insert_indexes.emplace_back(t, target[t]);
+                  FC_ASSERT(result.insert_indexes.size() <= MAX_NUM_ARRAY_ELEMENTS);
```

Comment:
Issue is that we insert *before* the FC_ASSERT. So the FC_ASSERT can be fine because `result.insert_indexes.size() == MAX_NUM_ARRAY_ELEMENTS`, and the next `emplace_back` will go over.

=> change the `FC_ASSERT` to `<` and move before the `emplace_back`.

This applies to the other locations as well.